### PR TITLE
Labels

### PR DESCRIPTION
**Pull request by mikedecosta**
Add labels support to bamboo

bamboo_labels

### CODE REVIEW

**Comment by mikedecosta** [comment]
It will probably be worth in the future to provide some kind of integration api endpoint that packages some of this configuration info together so we're making 1 call instead of 3+

File: src/main/java/com/mabl/ConfiguratorServlet.java [hunk 1]
```diff
@@ -47,13 +62,13 @@ private void doGetEnvironments(RestApiClient apiClient, HttpServletResponse resp
...
+    private void doGetLabels(RestApiClient apiClient, HttpServletResponse response) {
         try {
             GetApiKeyResult getApiKeyResult = apiClient.getApiKeyResult(apiClient.getRestApiKey());
-            GetApplicationsResult getApplicationsResult = apiClient.getApplicationsResult(getApiKeyResult.organization_id);
-            writeToWriter(response, objectMapper.writeValueAsString(getApplicationsResult.applications));
+            GetLabelsResult getLabelsResult = apiClient.getLabelsResult(getApiKeyResult.organization_id);
+            writeToWriter(response, objectMapper.writeValueAsString(getLabelsResult.labels));
         } catch (JsonProcessingException | RuntimeException e) {
-            log.error(String.format("Unexpected status returned from doGetApplications: Reason '%s'", e.getMessage()));
+            log.error(String.format("Unexpected status returned from doGetLabels: Reason '%s'", e.getMessage()));
```
Comment (twistedpair) [inline src/main/java/com/mabl/ConfiguratorServlet.java]: Do we want to pass the raw error object to the `log.error` call as the second arg?

That way we get the whole trace, unless we don't want that.

File: src/main/java/com/mabl/RestApiClient.java [hunk 2]
```diff
@@ -142,11 +152,12 @@ private RequestConfig getDefaultRequestConfig() {
     private AbstractHttpEntity getCreateDeplotmentPayloadEntity(
```
Comment (twistedpair) [inline src/main/java/com/mabl/RestApiClient.java]: Spelling.

File: src/main/java/com/mabl/RestApiClient.java [hunk 3]
```diff
@@ -76,9 +79,11 @@ public String getRestApiKey() {
     public CreateDeploymentResult createDeploymentEvent(
             final String environmentId,
             final String applicationId,
-            final CreateDeploymentProperties properties) {
+            final Set<String> planLabels,
```
Comment (twistedpair) [inline src/main/java/com/mabl/RestApiClient.java]: Just curious why we're passing `null` through here, rather than an empty list.

Would simplify upstream code, and if the API needs it to be null, it can be packaged at the end of the line in the payload create step.

File: src/main/resources/editCreateDeployment.ftl [hunk 4]
```diff
@@ -56,7 +76,8 @@ var mabl =
             {
                 // Loop through each of the results and append the option to the dropdown
                 AJS.$.each(result, function(k, v) {
-                    dropdown.append('<option value="' + v.id + '">' + v.name + '</option>');
+                    var id = v.id ? v.id : v.name;
```
Comment (twistedpair) [inline src/main/resources/editCreateDeployment.ftl]: Could we get a code comment explaining why we'd not have an `id` and need to use a `name` value?

File: src/test/java/com/mabl/MablTestConstants.java [hunk 5]
```diff
@@ -1,19 +1,36 @@
 package com.mabl;
 
+import org.jetbrains.annotations.Nullable;
```
Comment (twistedpair) [inline src/test/java/com/mabl/MablTestConstants.java]: Is that really in scope? Perhaps we should use the `javax` `@Nullable` ?

Comment (mikedecosta) [reply src/test/java/com/mabl/MablTestConstants.java]: It's included in the bamboo api: https://mvnrepository.com/artifact/org.jetbrains/annotations/13.0/usages

```
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ bamboo-plugin ---
[INFO] com.mabl.bamboo:bamboo-plugin:atlassian-plugin:0.1.0
[INFO] +- com.atlassian.bamboo:atlassian-bamboo-web:jar:6.3.2:provided
[INFO] |  +- com.atlassian.bamboo:atlassian-bamboo-core:jar:6.3.2:provided
[INFO] |  |  +- com.atlassian.bamboo:atlassian-bamboo-api:jar:6.3.2:provided
[INFO] |  |  |  +- com.atlassian.bamboo:atlassian-bamboo-api-agent-bootstrap:jar:6.3.2:provided
[INFO] |  |  |  +- com.atlassian.extras:atlassian-extras-api:jar:3.3.0:provided
[INFO] |  |  |  +- com.atlassian.config:atlassian-config:jar:0.22:provided
[INFO] |  |  |  +- com.atlassian.plugins:atlassian-plugins-servlet:jar:4.1.0:provided
[INFO] |  |  |  +- com.atlassian.plugins:atlassian-plugins-webfragment-api:jar:4.0.0:provided
[INFO] |  |  |  +- com.atlassian.bandana:atlassian-bandana:jar:3.1:provided
[INFO] |  |  |  +- com.atlassian.cache:atlassian-cache-api:jar:2.11.3:provided
[INFO] |  |  |  |  +- com.atlassian.bundles:jsr305:jar:1.1:provided
[INFO] |  |  |  |  \- com.atlassian.instrumentation:atlassian-instrumentation-core:jar:2.2.1:provided
[INFO] |  |  |  +- com.atlassian.bucket:atlassian-bucket:jar:5.1.4:provided
[INFO] |  |  |  +- org.apache.ant:ant:jar:1.8.4:provided
[INFO] |  |  |  |  \- org.apache.ant:ant-launcher:jar:1.8.4:provided
[INFO] |  |  |  +- org.apache.commons:commons-text:jar:1.1:provided
[INFO] |  |  |  +- commons-configuration:commons-configuration:jar:1.4-atlassian-1:provided
[INFO] |  |  |  |  \- commons-jxpath:commons-jxpath:jar:1.2:provided
[INFO] |  |  |  +- commons-beanutils:commons-beanutils:jar:1.9.2:provided
[INFO] |  |  |  +- opensymphony:osuser:jar:1.0-20060106:provided
[INFO] |  |  |  +- org.jetbrains:annotations:jar:13.0:provided
[INFO] |  |  |  +- net.jcip:jcip-annotations:jar:1.0:provided

```
